crates/driver/src/domain/competition/auction.rs

```diff
@@ -142,66 +143,56 @@ impl Auction {
-                        order::Partial::Yes { executed } => executed.0,
-                        order::Partial::No => 0.into(),
-                    },
-                    partially_fillable: match order.partial {
-                        order::Partial::Yes { .. } => true,
-                        order::Partial::No => false,
-                    },
-                })
-                .context("Remaining::from_order")?;
-                let sell = remaining
-                    .remaining(order.sell.amount.0)
-                    .context("remaining_sell")?;
-                let fee = remaining
-                    .remaining(order.fee.user.0)
-                    .context("remaining_fee")?;
-                sell.checked_add(fee).context("add sell and fee")
-            }
+                let sell = util::math::mul_ratio(order.sell.amount.0, available, target)?;
+                let fee = util::math::mul_ratio(order.fee.user.0, available, target)?;
 
-            let max_fill = match max_fill(order) {
-                Ok(balance) => balance,
-                Err(err) => {
-                    let reason =
-                        observe::OrderExcludedFromAuctionReason::CouldNotCalculateRemainingAmount(
-                            &err,
-                        );
-                    observe::order_excluded_from_auction(order, reason);
+                sell.checked_add(fee)
+            }
+            let max_sell = match max_sell(order) {
+                Some(amount) => amount,
+                None => {
+                    observe::order_excluded_from_auction(
+                        order,
+                        observe::OrderExcludedFromAuctionReason::CouldNotCalculateMaxSell,
+                    );
                     return false;
                 }
             };
 
-            let used_balance = match order.is_partial() {
-                true => {
-                    if *remaining_balance == 0.into() {
-                        return false;
-                    }
-                    max_fill.min(*remaining_balance)
-                }
-                false => {
-                    if *remaining_balance < max_fill {
-                        return false;
-                    }
-                    max_fill
-                }
+            let used_sell = match order.partial {
+                order::Partial::Yes { .. } => max_sell.min(*remaining_balance),
+                order::Partial::No if max_sell <= *remaining_balance => max_sell,
+                _ => U256::zero(),
             };
-            *remaining_balance -= used_balance;
+            if used_sell.is_zero() {
+                observe::order_excluded_from_auction(
+                    order,
+                    observe::OrderExcludedFromAuctionReason::InsufficientBalance,
+                );
+                return false;
+            }
+
+            if let order::Partial::Yes { available } = &mut order.partial {
+                available.0 =
```

Comment:
What is this doing? Haven't we already checked for the case where remaining fill amount of the order is 0 (above when we max_sell.min(remaining_balance))?

Reply:
I'm also confused. Wasn't the value in `Partial::Yes` already `min(user_balance, order.remaining_amount)` going into the function?
If not it seems like we initialize the `Partial::Yes` variant with `value == user_balance` and leave it with `min(user_balance, order.remaining_amount)`.
If this transformation is really what's happening here we should have 2 different enums for that.

Reply:
We need to scale available amounts to available balances. That is, if we have a partially fillable sell order that sells 100 with a fee of 10, which has already executed 20 (and 2 fee). This means that, before balance allocation 80 is “available”.

Now, once we do balance allocation (what this auction code does), it can be that the user only has a balance of 30. Here, we have 80 sell available meaning that we need `(80 + 8 /*fee*/)` sell tokens for the full order. This means that our available sell amount is actually only `80 * 30 / 88 = 27` so that there is enough room for the fee (we cannot use `available = 30` because otherwise there would not be enough tokens left over for the scaled fee).

This is complicated and took some thought - and **needs** testing. I’m trying to write a nice E2E test that takes an auction, with some users with balances and asserts that the orders that get passed to a solver match some expectations.

Reply:
> Wasn't the value in Partial::Yes already min(user_balance, order.remaining_amount) going into the function?

No. This is a bit of a code smell from our current auction setup in that we get use the same `order` type for auction that we receive from the `autopilot` and that we use in the `competition`. This means that things like:
- liquidity
- token information
- order balance information

Are not correctly populated until the auction is “prepared”.

I personally see this as a code smell, but I also think that fixing it is outside the scope of this PR. I will document that going in to the `retain_mut` function, balances have not been checked and allocated to the orders, so that is what is going on here.

Reply:
For me, I guess what I was missing was that this is now updating the order with the previously computed information (🤦). 
Then, the reason for why we cannot simply set available.0 to used sell (because of fee if remaining balance is smaller than max_sell) is non trivial and should IMO either be documented or be made more explicit by using an explicit if statement (and setting used_sell in the other cases).

Reply:
> be documented

https://github.com/cowprotocol/services/pull/1900#discussion_r1339732324

Reply:
> using an explicit if statement (and setting used_sell in the other cases).

This is also incorrect because they are different units. `avaialble` is a "target amount" (in buy token for buy orders, in sell token for sell orders), `used_sell` is a "sell amount".

I authored fd15cb145ee8f0eb628492d86703f2baa5013290 to use explicit typing and hopefully explain things better.